src/procedures/proc_common_neighbors.c

```diff
@@ -0,0 +1,316 @@
+	ctx->DEGREE = GrB_NULL;
+	ctx->select_op = GrB_NULL;
+	ctx->graph = QueryCtx_GetGraph();
+
+	// Setup reusable output.
+	ctx->output = array_new(SIValue, 6);
+	ctx->output = array_append(ctx->output, SI_ConstStringVal("node1"));
+	ctx->output = array_append(ctx->output, SI_Node(NULL)); // Place holder.
+	ctx->output = array_append(ctx->output, SI_ConstStringVal("node2"));
+	ctx->output = array_append(ctx->output, SI_Node(NULL)); // Place holder.
+	ctx->output = array_append(ctx->output, SI_ConstStringVal("score"));
+	ctx->output = array_append(ctx->output, SI_DoubleVal(0.0)); // Place holder.
+
+	return ctx;
+}
+
+// Free similarity context.
+static void SimilarityContext_Free(SimilarityContext *ctx) {
+	if(ctx) {
+		if(ctx->output) array_free(ctx->output);
+		if(ctx->At != GrB_NULL) GrB_free(&ctx->At);
+		if(ctx->SIM != GrB_NULL) GrB_free(&ctx->SIM);
+		if(ctx->desc != GrB_NULL) GrB_free(&ctx->desc);
+		if(ctx->thunk != GrB_NULL) GrB_free(&ctx->thunk);
+		if(ctx->DEGREE != GrB_NULL) GrB_free(&ctx->DEGREE);
+		if(ctx->select_op != GrB_NULL) GrB_free(&ctx->select_op);
+		if(ctx->iter) GxB_MatrixTupleIter_free(ctx->iter);
+		rm_free(ctx);
+	}
+}
+
+/* Selects the upper triangle portion of a matrix.
+ * taking into account row offset. */
+bool _select_triu(GrB_Index i, GrB_Index j, GrB_Index nrows, GrB_Index ncols, const void *x,
+				  const void *thunk) {
+
+	int64_t offset = *((int64_t *)thunk);
+	// Upper triangular.
+	return j > (offset + i);
+}
+
+// Extract row min through max from `A`, store result is `C`.
+static void _extract_rows(GrB_Matrix C, GrB_Matrix A, GrB_Index min, GrB_Index max) {
+	assert(C && A && (min <= max));
+
+	GrB_Info info;
+	GrB_Index nrows;
+	GrB_Index ncols;
+	GrB_Matrix_nrows(&nrows, A);
+	GrB_Matrix_ncols(&ncols, A);
+
+	max = MIN(max, nrows);
+
+	// Build array of row indices to extract.
+	GrB_Index count = max - min;
+	GrB_Index rows[count];
+	for(GrB_Index i = 0; i < count; i++) rows[i] = min + i;
+
+	info = GrB_Matrix_nrows(&nrows, C);
+	assert(info == GrB_SUCCESS);
+
+	// Make sure C dimensions matches.
+	if(nrows != count) {
+		info = GxB_Matrix_resize(C, count, ncols);
+		assert(info == GrB_SUCCESS);
+	}
+
+	// Extract row min through max.
+	info = GrB_Matrix_extract(C, GrB_NULL, GrB_NULL, A, rows, count, GrB_ALL, ncols, GrB_NULL);
+	assert(info == GrB_SUCCESS);
+}
+
+static void _reiterate(SimilarityContext *ctx) {
+	GrB_Info info;
+	GrB_Index nvals;
+
+	GrB_Index min = ctx->iteration * ctx->batch_size;
+	GrB_Index max = min + ctx->batch_size;
+	_extract_rows(ctx->SIM, ctx->A, min, max);
+
+	// Compute (f * A) * At, [A*At][i,j] = Number of paths between i and j.
+	info = GrB_mxm(ctx->SIM, GrB_NULL, GrB_NULL, GxB_PLUS_TIMES_INT64, ctx->SIM, ctx->At, GrB_NULL);
+	assert(info == GrB_SUCCESS);
+
+	info = GxB_Scalar_setElement_INT64(ctx->thunk, min);
+	assert(info == GrB_SUCCESS);
+
+	info = GxB_Matrix_select(ctx->SIM, GrB_NULL, GrB_NULL, ctx->select_op, ctx->SIM, ctx->thunk,
+							 ctx->desc);
```

Comment:
The built-in TRIU seems to work for me here:
`GxB_Matrix_select(ctx->SIM, GrB_NULL, GrB_NULL, GxB_TRIU, ctx->SIM, ctx->thunk, ctx->desc)`
Can we use it instead of a custom operator? The documentation indicates that a Thunk of 1 will skip the diagonal.

Reply:
We could have used that if we were handling the entire matrix at once, but due to our bulk processing
the elements which are at the upper triangle position for batch 2..n are shifted, the shift amount is specified by the thunk.

Reply:
I think it's still viable. The documentation indicates that the built-in TRIU will use the thunk as the shift amount:
> For the GxB_TRIL, GxB_TRIU, GxB_DIAG, and GxB_OFFDIAG, the Thunk
> parameter may be NULL, or it may be present but contain no entry. In this
> case, these operators use the value of k=0, the main diagonal. If present,
> the Thunk can be any built-in type. The value of this entry is typecasted:
> k = (int64_t) Thunk. The value k=0 specifies the main diagonal of the
> matrix, k=1 is the +1 diagonal (the entries just above the main diagonal),
> k=-1 is the -1 diagonal, and so on.

The comment at https://github.com/RedisGraph/RedisGraph/blob/master/deps/GraphBLAS/Include/GraphBLAS.h#L961 indicates that the operator will check that `((j-i) >= thunk)`, so I think we just need to add 1 to our current thunk to get the desired behavior.